### PR TITLE
Adds keep-alive to gRPC connections between the frontend and history/matching services

### PR DESCRIPTION
**Pull request by mastermanu**
This PR adds a 10-second keep-alive between the frontend and history/matching services. This will enable a request on a severed connection to fail more quickly so that it can be retried, therefore reducing recovery time if one of those services recycles.

This is a very low-risk change that is automatically validated via unit/integration testing

### CODE REVIEW

File: common/rpc/grpc.go [hunk 1]
```diff
@@ -91,6 +91,20 @@ func Dial(hostName string, tlsConfig *tls.Config, logger log.Logger) (*grpc.Clie
...
 		grpc.WithConnectParams(cp),
+	}
+
+	if enableKeepAlive {
+		dialOptions = append(dialOptions, grpc.WithKeepaliveParams(
+			keepalive.ClientParameters{
+				Time: 10 * time.Second,
+			},
+		),
+		)
```
Comment (wxing1292) [inline common/rpc/grpc.go:102]: ``))``

**Review by wxing1292 — APPROVED** [review]

File: common/rpc/grpc.go (hunk 1 above)
Comment (mastermanu) [reply common/rpc/grpc.go]: fixed

File: common/rpc/rpc.go [hunk 2]
```diff
@@ -204,7 +204,7 @@ func (d *RPCFactory) CreateFrontendGRPCConnection(hostName string) *grpc.ClientC
 		}
 	}
 
-	return d.dial(hostName, tlsClientConfig)
+	return d.dial(hostName, tlsClientConfig, false)
```
Comment (paulnpdev) [inline common/rpc/rpc.go:207]: Nit:
(newish to golang, don't hesitate to tell me I'm wrong here, but ...)
I'd have written this code, just for readability:

keepAlive := false
return d.dial(hostName, tlsClientConfig, keepAlive)
... or even
return d.dial(hostName, tlsClientConfig, false) // keepAlive=false

Personally I think Golang should have included named args; I think it's one of the things python got right.

File: common/rpc/grpc.go [hunk 3]
```diff
@@ -91,6 +91,19 @@ func Dial(hostName string, tlsConfig *tls.Config, logger log.Logger) (*grpc.Clie
 		grpc.WithDefaultServiceConfig(DefaultServiceConfig),
 		grpc.WithDisableServiceConfig(),
 		grpc.WithConnectParams(cp),
+	}
+
+	if enableKeepAlive {
+		dialOptions = append(dialOptions, grpc.WithKeepaliveParams(
+			keepalive.ClientParameters{
+				Time: 10 * time.Second,
```
Comment (paulnpdev) [inline common/rpc/grpc.go:99]: Shouldn't this value (10 seconds) be added to the constants defined at line 45?

File: common/rpc/rpc.go [hunk 4]
```diff
@@ -218,11 +218,11 @@ func (d *RPCFactory) CreateInternodeGRPCConnection(hostName string) *grpc.Client
 		}
 	}
 
-	return d.dial(hostName, tlsClientConfig)
+	return d.dial(hostName, tlsClientConfig, true)
```
Comment (paulnpdev) [inline common/rpc/rpc.go:221]: ditto, here and other places in CL

**Pull request merged**